### ISSUE TITLE
Improved lossless audio support

### ISSUE THREAD

**Issue by OxygenCobalt** [OxygenCobalt/Auxio#174]
Auxio's lossless audio support is somewhat lacking. This is caused by several issues:
1. ExoPlayer resamples very high quality 24-bit or 32-bit PCM audio down to 16-bit PCM audio. This kills any quality that a loss-less audio file might have.
2. Worse still, the resampling operation and my own ReplayGain engine adds enough overhead as to cause distorted, choppy playback in certain situations, as is covered by #151 and (maybe) #157.

This issue is about some internal technical reworks to fix this. The great thing about ExoPlayer is that you can rip open the player implementation and replace nearly everything with your own code, and that is (to an extent) what I want to do.

Basically, this is my plan (in technical terms):
1. Rework the ReplayGain system to handle far more encodings and use the lower level audio processor primitives. This will allow me to avoid a copy when ReplayGain does not need to be applied.
2. Enable float output in the audio sink. Not sure what implications this might have for performance, but I hope it's better than the current resampling strategy.
3. Override the audio sink to remove the default audio processor chain and replace it with my own. This will be solely composed of the ReplayGain system.
4. (Not confirmed) If I implement an equalizer or playback speed support, additional audio processors will need to be implemented that need to be similarly low-level.

**Comment by KraXen72** [comment]
this is all super cool, but like, do you just *know* how exoplayer and streaming audio on android works? this all seems like you've been working with android for 20+ years lol. I'm amazed and keep up the good work!

**Comment by OxygenCobalt** [comment]
Yeah I basically know the internals of android's media system, however I that's the product of 2 years of developing this app and going "No, that solution isn't good enough, I'm going to find a better way." I've also developed a good habit of not completely relying on external documentation and actually reading the source code of what I use. `grep` and `cs.android.com` are life-savers in that regard.

**Comment by KraXen72** [comment]
Oh, really nice. respect. thanks for answering.

**Comment by dreamsyntax** [comment]
Was going to take a look at this, but see the next-release tag. Has this already been completed in a wip branch not yet pushed?

**Comment by OxygenCobalt** [comment]
`next-release` signifies not that I've completed it, but rather that I'm aiming to work on it for the next major release. In this case, I'm aiming for this, #108, #93, and #85 to be added to 2.6.1/2.7.0.

**Comment by OxygenCobalt** [comment]
Okay, things have gotten more difficult. I was under the impression that I could side-step ExoPlayer's default float output behavior, which disables all audio manipulation like ReplayGain. Turns out I psysically *cannot* do it with the public AudioSink API.

Thus, I am left with a few options:
1. Make the better lossless audio functionality a setting. I cannot do this, as it is impossible to enable float output or swap out AudioSink implementations at runtime.
2. Reflect and replace the audio processors myself. This is probably not a good idea.
3. Talk to the ExoPlayer devs about why audio processors are disabled with float output and if it is safe for me to side-step it on my own end. This is probably what I'm going to try doing.

**Comment by OxygenCobalt** [comment]
Okay, I made a related issue [here](https://github.com/google/ExoPlayer/issues/10516). I'll try to start work on a PR eventually.

**Comment by OxygenCobalt** [comment]
Started the first PR in what's probably going to be a series of changes I'll make to ExoPlayer eventually [here](https://github.com/google/ExoPlayer/pull/10565). This should allow me to *use* float audio, but trying to optimize the resampling problem is much harder.

**Comment by OxygenCobalt** [comment]
Actually, I'm not even going to try to upstream this. I think I'm just going to maintain my own ExoPlayer fork with these fixes hacked in for now, any maybe I'll refine it enough in the future so that it's mergable.

**Comment by OxygenCobalt** [comment]
Okay, so I can feasibly implement this for the next release. I just don't want to. I've had to mangle so many components that won't play along with non 16-bit PCM audio and I am extremely worried that knock-on effects that might have. I need to put more effort into my custom ExoPlayer fork and rework at least some of the fundamental components just to be safe, and even then. I'm going to shift priority, I think.

**Comment by RustoMCSpit** [comment]
mark this as off topic. do you contribute to exoplayer when you make massive overhauls like this or just downstream here in auxio?

**Comment by OxygenCobalt** [comment]
> mark this as off topic. do you contribute to exoplayer when you make massive overhauls like this or just downstream here in auxio?

Yes. There are some patches I need to backport but I don't have the bandwidth right now to do it. It's also likely that at least with this issue that I would probably cowboy this functionality into not-instantly-failing and then make it workable as a PR later on @RustoMCSpit

**Comment by Switch123456789** [comment]
As mentioned in another report, I use Wavpack for all my music (which is widely unsupported on Android). So as a workaround I created flac-conversions just for my phone. These files were created with ffmpeg and have no tags or anything like that.

I noticed that every 32-bit file doesn't playback and ultimately doesn't let Auxio play any other files after that until I clear the whole queue. According to your post, it should work (with bad quality). So I am wondering whats up with that?  Let me know if I can provide more info to help!

**Comment by OxygenCobalt** [comment]
>I noticed that every 32-bit file doesn't playback and ultimately doesn't let Auxio play any other files after that until I clear the whole queue. According to your post, it should work (with bad quality). So I am wondering whats up with that? Let me know if I can provide more info to help!

Is it making the app unresponsive? That's incredibly odd albeit the way I "handle errors" (read: skip to the next song cuz when I wrote that code I was 15 and error dialogs sounded hard) is really bad and could be making things far worse @Switch123456789

**Comment by Switch123456789** [comment]
> Is it making the app unresponsive?

The interface keeps working fine. To be precise, if playback of a 32 bit file is requested, the interface shows: 0:00 progress and there is no audio. If I skip right away, it will continue playing the other files. However, if I wait for 5-10 seconds before I skip the 32bit file, it will not load/play any files from the queue anymore until it's cleared.

**Comment by OxygenCobalt** [comment]
That's such a weird issue. Implies that I think your phone's codec handler is not working well. What ROM do you use @Switch123456789?

**Comment by Switch123456789** [comment]
I am on [iode-7.5-20260427-FP6](https://gitlab.iode.tech/ota/betas/-/blob/master/iode-7.5-20260427-FP6-ota.zip.sha256), it's based on LOS23.2/A16. The weird thing is, VLC can play some files with stuttering sound but the stock music app can play all files without any issues. Auxio behaves differently with different 32bit formatted files but is never able to play them back.
